cpp/src/lance/encodings/binary.h

```diff
@@ -95,6 +103,72 @@ ::arrow::Result<std::shared_ptr<::arrow::Scalar>> VarBinaryDecoder<T>::GetScalar
   return std::make_shared<typename ::arrow::TypeTraits<T>::ScalarType>(buf);
 }
 
+template <ArrowType T>
+::arrow::Result<std::shared_ptr<::arrow::Int64Array>> VarBinaryDecoder<T>::ReadPositions(
+    int32_t start, int32_t length) const {
+  auto buf = infile_->ReadAt(position_ + start * sizeof(int64_t), (length + 1) * sizeof(int64_t));
+  if (!buf.ok()) {
+    return ::arrow::Status::IOError(fmt::format(
+        "VarBinaryDecoder::ReadPositions: failed to read positions: start={}, length={}: {}",
+        start,
+        length,
+        buf.status().message()));
+  }
+  return std::make_shared<typename ::arrow::TypeTraits<OffsetType>::ArrayType>(length + 1, *buf);
+}
+
+template <ArrowType T>
+::arrow::Result<std::shared_ptr<::arrow::Array>> VarBinaryDecoder<T>::Take(
+    std::shared_ptr<::arrow::Int32Array> indices) const {
+  if (indices->length() == 0) {
+    return MakeEmpty();
+  }
+  auto start = indices->Value(0);
+  auto last = indices->Value(indices->length() - 1);
+  auto length = last - start + 1;
+  ARROW_ASSIGN_OR_RAISE(auto positions, ReadPositions(start, length));
+  assert(positions->length() == length + 1);
+
+  const int64_t kMinimalBatchBytes = 128 * 1024;  // 128K
```

Comment:
does this need to be enforced elsewhere? if so, do we need a global constant?

Reply:
this is an optimization, does not change behavior. 

Long term this could be a runtime configuration tho.